server/telemetry.go

```diff
@@ -68,3 +94,53 @@ func (p *Plugin) initTelemetry(enableDiagnostics *bool) error {
 	}
 	return nil
 }
+
+func (p *Plugin) handleTrackEvent(w http.ResponseWriter, r *http.Request) {
+	var res httpResponse
+	defer p.httpAudit("handleTrackEvent", &res, w, r)
+
+	p.mut.RLock()
+	telemetryEnabled := p.telemetry != nil
+	p.mut.RUnlock()
+
+	if !telemetryEnabled {
+		res.Err = "telemetry is disabled"
+		res.Code = http.StatusBadRequest
+		return
+	}
+
+	var data trackEventRequest
+	if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, requestBodyMaxSizeBytes)).Decode(&data); err != nil {
+		res.Err = err.Error()
+		res.Code = http.StatusBadRequest
+		return
+	}
+
+	if _, ok := telemetryClientEvents[data.Event]; !ok {
+		res.Err = "invalid telemetry event"
+		res.Code = http.StatusBadRequest
+		return
+	}
+
+	if _, ok := telemetryClientTypes[data.ClientType]; !ok {
+		res.Err = "invalid client type"
+		res.Code = http.StatusBadRequest
+		return
+	}
```

Comment:
I can maybe understand making sure the client type is correct, but I wonder if the event types needs to be strictly checked like this? Why not just send it through? Then there's one less place we need to update when we add an event.

Reply:
Funnily I've even considered strong typing the whole props set but it was just too hard to lose the flexibility as I couldn't possibly predict what could be useful there.

As for event types I still think the trade-off is worth it as I prefer having a single source of truth on the server that explicitly enumerates what's possible to send through the API. I've actually refactored a bit to make it slightly easier and less error prone.